src/Sulu/Bundle/MediaBundle/EventListener/MediaSecurityListener.php

```diff
@@ -0,0 +1,84 @@
+<?php
+
+/*
+ * This file is part of Sulu.
+ *
+ * (c) Sulu GmbH
+ *
+ * This source file is subject to the MIT license that is bundled
+ * with this source code in the file LICENSE.
+ */
+
+namespace Sulu\Bundle\MediaBundle\EventListener;
+
+use Sulu\Bundle\MediaBundle\Admin\MediaAdmin;
+use Sulu\Bundle\MediaBundle\Entity\Collection;
+use Sulu\Bundle\MediaBundle\Entity\CollectionRepositoryInterface;
+use Sulu\Component\Security\Authorization\PermissionTypes;
+use Sulu\Component\Security\Authorization\SecurityCheckerInterface;
+use Sulu\Component\Security\Authorization\SecurityCondition;
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+use Symfony\Component\HttpKernel\Event\RequestEvent;
+use Symfony\Component\HttpKernel\Exception\NotFoundHttpException;
+
+/**
+ * @internal this class is for internal usages
+ */
+class MediaSecurityListener implements EventSubscriberInterface
+{
+    /**
+     * @var CollectionRepositoryInterface
+     */
+    private $collectionRepository;
+
+    /**
+     * @var SecurityCheckerInterface
+     */
+    private $securityChecker;
+
+    public function __construct(
+        CollectionRepositoryInterface $collectionRepository,
+        SecurityCheckerInterface $securityChecker
+    ) {
+        $this->collectionRepository = $collectionRepository;
+        $this->securityChecker = $securityChecker;
+    }
+
+    public static function getSubscribedEvents(): array
+    {
+        return [
+            RequestEvent::class => 'onKernelRequest',
+        ];
+    }
+
+    public function onKernelRequest(RequestEvent $event): void
+    {
+        $request = $event->getRequest();
+
+        if (!$event->isMasterRequest()) {
+            return;
+        }
+
+        $routeName = $request->attributes->get('_route');
+
+        if (!\in_array($routeName, ['sulu_media.website.media.download', 'sulu_media.website.media.download_admin'])) {
+            return;
+        }
+
+        $collectionId = $this->collectionRepository->findIdByMediaId((int) $request->attributes->get('id'));
+
+        if (!$collectionId) {
+            throw new NotFoundHttpException();
```

Comment:
As the media can not exist without a collection a NotFoundHttpException is correct here, as its common that a old media url could still exist in some http cache or email. And if that media not exist it should return 404.

The other way would be that we don't throw any exception and just do a early return to avoid security check, then it would land in the MediaController but as the Media does not exist it will also correctly there throw a 404 Exception.